### PR TITLE
#5051 Add Array / Array Index Support to Var Analysis

### LINKED ISSUE #5051: Add Array / Array Index Support
Separating out add Array \/ Array Index support from <https://github.com/pixiebrix/pixiebrix-extension/issues/4591>\. <span data-preserve-white-space></span>

<p></p>

This will be the last issue for Autocomplete\/Autosuggest Expressions \(variable popover\) support "Phase 1"\.

<p></p>

### PR DESCRIPTION
**Pull request by BALEHOK**
## What does this PR do?

- Closes #5051

## Demo

https://www.loom.com/share/44375a19c71140d7a5f83c7fe99b7048

## Checklist

- [x] Add tests
- [x] Designate a primary reviewer @BLoe

### CODE REVIEW

File: src/runtime/pathHelpers.ts [hunk 1]
```diff
@@ -155,6 +155,10 @@ export function getPathFromArray(parts: Array<string | number>): string {
         return `["${part}"]`;
       }
 
+      if (typeof part === "number" || /^\d+$/.test(part)) {
+        return `[${part}]`;
```
Comment (BALEHOK) [inline src/runtime/pathHelpers.ts:159]: Nunjucks only supports access to array items in the format `array[n].property`

Comment (twschiller) [reply src/runtime/pathHelpers.ts:159]: This still works for variable inputs, correct?

File: src/analysis/analysisVisitors/varAnalysis/varAnalysis.ts [hunk 2]
```diff
@@ -58,10 +58,10 @@ async function setServiceVars(extension: FormState, contextVars: VarMap) {
   for (const service of extension.services ?? []) {
     // eslint-disable-next-line no-await-in-loop
     const serviceContext = await makeServiceContext([service]);
-    contextVars.setExistenceFromValues(
-      `${KnownSources.SERVICE}:${service.id}`,
-      serviceContext
-    );
+    contextVars.setExistenceFromValues({
```
Comment (twschiller) [inline src/analysis/analysisVisitors/varAnalysis/varAnalysis.ts:61]: 👍

File: src/analysis/analysisVisitors/varAnalysis/varMap.test.ts [hunk 3]
```diff
@@ -118,18 +123,118 @@ describe("setExistence", () => {
...
+    varMap.setExistence({
+      source: "brick1",
+      path: "@foo",
+      existence: VarExistence.DEFINITELY,
+      isArray: true,
+    });
+
+    varMap.setExistence({
+      source: "brick1",
+      path: "@foo.bar",
```
[the remark this answers is not in the document]
Comment (twschiller) [reply src/analysis/analysisVisitors/varAnalysis/varMap.test.ts:198]: Where this also comes into play is that from a variable analysis perspective, IIRC people can do `foo.length` in Nunjucks to get the length of the array. 

In the future, for the native JS types, we'll have to add those native JS properties to the variable existence map

File: src/analysis/analysisVisitors/varAnalysis/varMap.ts [hunk 4]
```diff
@@ -25,71 +90,73 @@ export enum VarExistence {
...
+
+  /**
+   * Whether the current node allows any child (i.e. doesn't have  a strict schema)
+   */
   [ALLOW_ANY_CHILD]?: boolean;
 
+  /**
+   * Whether the current node is an array (allows only numeric keys)
+   */
+  [IS_ARRAY]?: boolean;
```
Comment (twschiller) [inline src/analysis/analysisVisitors/varAnalysis/varMap.ts:108]: As mentioned here: https://github.com/pixiebrix/pixiebrix-extension/pull/5070/files#r1085563946, this flag also control the `path` format. As the IS_ARRAY implies another level of access before hitting the element types

**Review by twschiller — APPROVED** [review]

File: src/runtime/pathHelpers.ts (hunk 1 above)
Comment (BALEHOK) [reply src/runtime/pathHelpers.ts:159]: yes, it does

**Pull request merged**